lib/puppet/type/file.rb

```diff
@@ -799,12 +805,12 @@ def should_be_file?
     return true if self[:ensure] == :file
 
     # I.e., it's set to something like "directory"
-    return false if e = self[:ensure] and e != :present
+    return false if self[:ensure] && self[:ensure] != :present
 
     # The user doesn't really care, apparently
     if self[:ensure] == :present
-      return true unless s = stat
-      return(s.ftype == "file" ? true : false)
+      return true unless stat
+      return(stat.ftype == "file" ? true : false)
```

Comment:
This will end up `stat`ing the filesystem an extra time. Better to memorize what `stat` returns.

Reply:
looks like the `stat` method itself memoizes what it does down on line 830, and the cache should not be invalidated between the two calls to `stat` here